File: jar/src/main/kotlin/jline/lang/processes/Binomial.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.processes;

import static jline.GlobalConstants.Inf;

import jline.util.Maths;
import jline.util.Pair;
import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;
import org.apache.commons.math3.util.FastMath;

import java.io.Serializable;
import java.util.Random;

import static org.apache.commons.math3.special.Beta.regularizedBeta;

/**
 * A Binomial distribution
 */
public class Binomial extends DiscreteDistribution implements Serializable {

    public Binomial(int n, double prob) {
        super("Binomial", 2, new Pair<Double, Double>(0.0, Inf));
        this.setParam(1, "n", n); // number of trials
        this.setParam(2, "p", prob); // success probability
    }

    public double evalCDF(int t) {
        return evalCDF((double) t);
    }

    public double evalCDF(double t) {
        int n = (int) this.getParam(1).getValue();
        double p = (double) this.getParam(2).getValue();
        double ret = 0.0;
        if (t >= n) {
            ret = 1.0;
        } else if (t >= 0) {
            ret = 1.0 - regularizedBeta(p, t + 1.0, (n - t));
        }
        return ret;
    }

    public double evalPMF(int k) {
        int n = (int) this.getParam(1).getValue();
        double p = (double) this.getParam(2).getValue();
        return FastMath.exp(Maths.factln(n) - Maths.factln(k) - Maths.factln(n - k)) * FastMath.pow(p, k) * FastMath.pow(1 - p, n - k);
    }

    /**
     * Evaluates the Laplace-Stieltjes Transform at s.
     * For Binomial(n, p), LST(s) = (1 - p + p*e^(-s))^n
     *
     * @param s the Laplace domain variable
     * @return the LST value at s
     */
    public double evalLST(double s) {
        int n = (int) this.getParam(1).getValue();
        double p = (double) this.getParam(2).getValue();
        return FastMath.pow(1 - p + p * FastMath.exp(-s), n);
    }

    public double getMean() {
        return ((double) this.getParam(2).getValue()) * ((int) this.getParam(1).getValue());
    }

    public double getSCV() {
        return getVar() / getMean() / getMean();
    }

    public double getSkewness() {
        int n = (int) this.getParam(1).getValue();
        double p = (double) this.getParam(2).getValue();
        double q = 1 - p;

        return (q - p) / (Math.sqrt(n * p * q));
    }

    public double getVar() {
        int n = (int) this.getParam(1).getValue();
        double p = (double) this.getParam(2).getValue();
        return n * p * (1.0 - p);
    }

    @Override
    public double[] sample(int nsamples, Random random) {
        double[] ret = new double[(int) nsamples];
        int n = (int) this.getParam(1).getValue();
        double p = (double) this.getParam(2).getValue();
        for (int k = 0; k < nsamples; k++) {
            int acc = 0;
            for (int i = 0; i < n; i++) {
                if (random.nextDouble() <= p) {
                    acc++;
                }
            }
            ret[k] = acc;
        }
        return ret;
    }

    public MatrixCell getProcess() {
        MatrixCell representation = new MatrixCell();
        representation.set(0, Matrix.singleton(getMean()));
        representation.set(1, Matrix.singleton(getSCV()));
        return representation;
    }
}

File: matlab/src/lang/processes/Binomial.m
classdef Binomial < DiscreteDistribution
    % A Binomial distribution
    %
    % Copyright (c) 2018-2022, Imperial College London
    % All rights reserved.

    methods
        function self = Binomial(n, p)
            % SELF = BINOMIAL(N, P)

            self@DiscreteDistribution('Binomial',2,[0,Inf]);
            setParam(self, 1, 'n', n);
            setParam(self, 2, 'p', p);
        end

        function ex = getMean(self)
            % EX = GETMEAN()

            % Get distribution mean
            n = self.getParam(1).paramValue;
            p = self.getParam(2).paramValue;
            ex = p*n;
        end

        function V = getVar(self)
            % V = GETVAR()
            n = self.getParam(1).paramValue;
            p = self.getParam(2).paramValue;
            V = n*p*(1-p);
        end

        function SCV = getSCV(self)
            % SCV = GETSCV()

            % Get distribution squared coefficient of variation (SCV = variance / mean^2)
            p = self.getParam(2).paramValue;
            n = self.getParam(1).paramValue;
            SCV = (1-p) / (n*p);
        end

        function X = sample(self, nsamples)
            % X = SAMPLE(N)

            X = zeros(1, nsamples);
            p = self.getParam(2).paramValue;
            n = self.getParam(1).paramValue;

            for k = 1:nsamples
                acc = 0;
                for i = 1:n
                    if rand <= p
                        acc = acc + 1;
                    end
                end
                X(k) = acc;
            end
        end

        function Ft = evalCDF(self,t)
            % FT = EVALCDF(SELF,T)

            % Evaluate the cumulative distribution function at T

            p = self.getParam(2).paramValue;
            n = self.getParam(1).paramValue;
            Ft = 0.0;
            if t >= n
                Ft = 1.0;
            elseif t >= 0
                Ft = 1.0 - betainc(p, t + 1.0, n - t);
            end
        end

        function L = evalLST(self, s)
            % L = EVALST(S)
            % Evaluate the Laplace-Stieltjes transform of the distribution function at s
            % For Binomial(n, p), LST(s) = (1 - p + p*e^(-s))^n

            p = self.getParam(2).paramValue;
            n = self.getParam(1).paramValue;
            L = (1 - p + p * exp(-s))^n;
        end

        function P = evalPMF(self, k)
            % P = EVALPMF(K)

            % Evaluate the probability mass function at k            

            p = self.getParam(2).paramValue;
            n = self.getParam(1).paramValue;
            P = binopdf(k, n, p);
        end
        
        function proc = getProcess(self)
            % PROC = GETPROCESS()
            
            % Get process representation for non-Markovian distribution
            % Returns [mean, SCV] pair for use in network analysis
            proc = [self.getMean(), self.getSCV()];
        end
    end

end

